src/plugins/camera/camera_impl.cpp

```diff
@@ -1583,4 +1612,117 @@ void CameraImpl::format_storage_async(Camera::ResultCallback callback)
+    auto ret = prom->get_future();
+
+    list_photos_async(
+        photos_range, [prom](Camera::Result result, std::vector<Camera::CaptureInfo> photo_list) {
+            prom->set_value(std::make_pair(result, photo_list));
+        });
+
+    return ret.get();
+}
+
+void CameraImpl::list_photos_async(
+    Camera::PhotosRange photos_range, const Camera::ListPhotosCallback callback)
+{
+    {
+        std::lock_guard<std::mutex> lock(_status.mutex);
+
+        if (_status.is_fetching_photos) {
+            _parent->call_user_callback([callback]() {
+                callback(Camera::Result::Busy, std::vector<Camera::CaptureInfo>{});
+            });
+            return;
+        } else {
+            _status.is_fetching_photos = true;
+        }
+
+        if (_status.image_count == -1) {
+            LogErr() << "Cannot list photos: camera status has not been received yet!";
+            _parent->call_user_callback([callback]() {
+                callback(Camera::Result::Error, std::vector<Camera::CaptureInfo>{});
+            });
+            return;
+        }
+    }
+
+    int start_index;
+
+    switch (photos_range) {
+        case Camera::PhotosRange::SinceConnection:
+            start_index = _status.image_count_at_connection;
+            break;
+        case Camera::PhotosRange::All:
+        default:
+            start_index = 0;
+            break;
+    }
+
+    std::thread([this, start_index, callback]() {
+        std::unique_lock<std::mutex> lock(_captured_request_mutex);
+
+        for (int i = start_index; i < _status.image_count; i++) {
+            // In case the vehicle sends capture info, but not those we are asking, we do not
+            // want to loop infinitely. The safety_count is here to abort if this happens.
+            auto safety_count = 0;
+            const auto safety_count_boundary = 3;
+
+            while (_status.photo_list.find(i) == _status.photo_list.end() &&
+                   safety_count < safety_count_boundary) {
+                safety_count++;
+
+                auto request_try_number = 0;
+                const auto request_try_limit =
+                    3; // Timeout if the request times out that many times
+                auto cv_status = std::cv_status::timeout;
+
+                while (cv_status == std::cv_status::timeout) {
+                    request_try_number++;
+                    if (request_try_number >= request_try_limit) {
+                        _parent->call_user_callback([callback]() {
+                            callback(Camera::Result::Timeout, std::vector<Camera::CaptureInfo>{});
+                        });
+                        return;
+                    }
+
+                    _parent->send_command_async(
+                        make_command_request_camera_image_captured(i), nullptr);
+                    cv_status =
+                        _captured_request_cv.wait_for(lock, std::chrono::duration<double>(1));
```

Comment:
I timeout after 1 second here, that's completely arbitrary. @julianoes is it an okay-ish value?